crypto/cversion.c

```diff
@@ -44,8 +48,41 @@ const char *OPENSSL_version_build_metadata(void)
 
 extern char ossl_cpu_info_str[];
 
+#if defined(_WIN32) && defined(OSSL_WINCTX)
+/* size: MAX_PATH + sizeof("OPENSSLDIR: \"\"") */
+static char openssldir[MAX_PATH + 15];
+
+/* size: MAX_PATH + sizeof("ENGINESDIR: \"\"") */
+static char enginesdir[MAX_PATH + 15];
+
+/* size: MAX_PATH + sizeof("MODULESDIR: \"\"") */
+static char modulesdir[MAX_PATH + 15];
+
+static CRYPTO_ONCE version_strings_once = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(version_strings_setup)
+{
+    BIO_snprintf(openssldir, sizeof(openssldir), "OPENSSLDIR: \"%s\"",
+                 ossl_get_openssldir());
+    BIO_snprintf(enginesdir, sizeof(enginesdir), "ENGINESDIR: \"%s\"",
+                 ossl_get_enginesdir());
+    BIO_snprintf(modulesdir, sizeof(modulesdir), "MODULESDIR: \"%s\"",
+                 ossl_get_modulesdir());
+    return 1;
+}
+
+# define TOSTR(x) #x
+# define OSSL_WINCTX_STRING "OSSL_WINCTX: \"" ## TOSTR(OSSL_WINCTX) ## "\""
+
+#endif
```

Comment:
We do we have this for Windows only?  I leads to a hard to read set of cases in `OpenSSL_version()`, which could be simplified to rely on these strings at all times, and thereby make the output robust to future changes that's made in one spot but forgotten in the other spot.

Reply:
OSSL_WINCTX_STRING would of course have to be conditional:

``` C
# ifdef OSSL_WINCTX
#  define OSSL_WINCTX_STRING "OSSL_WINCTX: \"" ## TOSTR(OSSL_WINCTX) ## "\""
# else
#  define OSSL_WINCTX_STRING "OSSL_WINCTX: Undefined"
# endif

Reply:
Because only on Windows there is this MAX_PATH limitation. There is no such thing on other platforms and I would have to allocate strings or do other tricks which I do not want to do.

Reply:
Hmm, ok